### PR TITLE
First cut at gnatmetrics tool training

### PR DESCRIPTION
**Pull request by frank-at-adacore**
Need some help on getting coupling metrics to actually work before we're ready to go

### CODE REVIEW

File: courses/gnatmetric/010_gnatmetric.rst [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
++ Given a control flow graph of a program
+
+  + E - number of edges
+  + N - number of nodes
+  + P - number of connected components (exit nodes)
+
++ The complexity is computed by:
+
+  + E - N + 2 * P
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:188]: Use latex equation there

File: courses/gnatmetric/010_gnatmetric.rst [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    Maximum depth of nested loops
+
+**Statement Complexity**
+    Complexity introduced by control statements only, without taking into account short-circuit forms 
+
+---------------------------------
+Understanding McCabe Complexity
+---------------------------------
+
+http://www.mccabe.com/pdf/mccabe-nist235r.pdf
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:178]: Missing `:url:` decorator

File: courses/gnatmetric/010_gnatmetric.rst [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+     All units making up a program
+
+  **Control coupling**
+     Dependencies between unit and other units that contain subprograms
+
+------------------------------
+Coupling Metrics
+------------------------------
+
++ Uses Ada's approach to definition of *class*
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:242]: ARM definiton of class is

> A class of types is a set of types that is closed under derivation

That is, the case for most Ada type, e.g. `Integer` and its children, which are not included in remaining bullet points.
Did you mean "Based on Ada's class-wide polymorphism" ?

File: courses/gnatmetric/010_gnatmetric.rst [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+  + Tagged types declared within packages
+  + Interface types declared within packages
+
++ Two kinds of coupling computed
+
+  **Fan-out coupling**
+    Number of classes given class
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:250]: Missing a verb, number of classes the given class depends on.

File: courses/gnatmetric/010_gnatmetric.rst [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+Select body file to get results for that particular file
+
+.. image:: ../../images/gnatmetric/body_results.jpg
+
+------------------------
+Line Metrics Explained
+------------------------
+
+**Average Lines In Body**
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:100]: Nit: those are definitions, for which we do have a `:dfn:` decorator now, if this decorator is a bit too "bright" to be used repeatedly we could reduce its brightness and replace it by a brighter `:term:` when needed (terms are included in a glossary in Sphinx, something cool to think about for our own use :) )

File: courses/gnatmetric/010_gnatmetric.rst [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+     Classes in traditional object-oriented sense
+
+  **Unit coupling**
+     All units making up a program
+
+  **Control coupling**
+     Dependencies between unit and other units that contain subprograms
+
+------------------------------
+Coupling Metrics
```
Comment (leogermond) [inline courses/gnatmetric/010_gnatmetric.rst:239]: I believe it is talking about OO coupling metrics only?

**Review by leogermond — CHANGES_REQUESTED** [review]
Thanks, I think I found some typos. Having some examples would be great too, I'm thinking auto-generated output lines from pieces of code, which is possible since GNATmetric is part of the CE

File: courses/gnatmetric/010_gnatmetric.rst (hunk 3 above)
Comment (frank-at-adacore) [reply courses/gnatmetric/010_gnatmetric.rst:242]: Added indication that only polymorphic classes are measured

**Review by leogermond — APPROVED** [review]

**Pull request merged**